File: packages/companion_protocol/lib/src/companion_crypto.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:crypto/crypto.dart' as hashes;
import 'package:cryptography/cryptography.dart';

import 'companion_models.dart';
import 'companion_pairing.dart';

class CompanionIdentity {
  CompanionIdentity._({
    required this.privateSeed,
    required this.publicKey,
    required this.fingerprint,
  });

  final List<int> privateSeed;
  final SimplePublicKey publicKey;
  final String fingerprint;

  static Future<CompanionIdentity> generate() async {
    final algorithm = Ed25519();
    final pair = await algorithm.newKeyPair();
    final seed = await pair.extractPrivateKeyBytes();
    final publicKey = await pair.extractPublicKey();
    return CompanionIdentity._(
      privateSeed: List<int>.unmodifiable(seed),
      publicKey: publicKey,
      fingerprint: companionFingerprint(publicKey.bytes),
    );
  }

  static Future<CompanionIdentity> fromSeed(List<int> seed) async {
    if (seed.length != 32) {
      throw const CompanionProtocolException(
        'INVALID_IDENTITY_SEED',
        'Identity seed must be 32 bytes.',
      );
    }
    final pair = await Ed25519().newKeyPairFromSeed(seed);
    final publicKey = await pair.extractPublicKey();
    return CompanionIdentity._(
      privateSeed: List<int>.unmodifiable(seed),
      publicKey: publicKey,
      fingerprint: companionFingerprint(publicKey.bytes),
    );
  }

  Future<Signature> sign(List<int> bytes) async {
    final pair = await Ed25519().newKeyPairFromSeed(privateSeed);
    return Ed25519().sign(bytes, keyPair: pair);
  }
}

class CompanionSignedPairing {
  const CompanionSignedPairing({
    required this.transcript,
    required this.initiatorPublicKey,
    required this.initiatorSignature,
    required this.responderPublicKey,
    required this.responderSignature,
  });

  final CompanionPairingTranscript transcript;
  final SimplePublicKey initiatorPublicKey;
  final Signature initiatorSignature;
  final SimplePublicKey responderPublicKey;
  final Signature responderSignature;

  Future<void> verify() async {
    if (companionFingerprint(initiatorPublicKey.bytes) !=
            transcript.initiatorFingerprint ||
        companionFingerprint(responderPublicKey.bytes) !=
            transcript.responderFingerprint) {
      throw const CompanionProtocolException(
        'PAIRING_FINGERPRINT_MISMATCH',
        'Pairing identity fingerprint does not match the transcript.',
      );
    }
    final bytes = transcript.canonicalBytes();
    final validInitiator = await Ed25519().verify(
      bytes,
      signature: initiatorSignature,
    );
    final validResponder = await Ed25519().verify(
      bytes,
      signature: responderSignature,
    );
    if (!validInitiator || !validResponder) {
      throw const CompanionProtocolException(
        'PAIRING_SIGNATURE_INVALID',
        'Pairing identity signature is invalid.',
      );
    }
  }
}

class CompanionSession {
  CompanionSession._({
    required this.sessionId,
    required SecretKey sendKey,
    required SecretKey receiveKey,
    required int sendDirection,
    required int receiveDirection,
    required int expiresAtMs,
  }) : _sendKey = sendKey,
       _receiveKey = receiveKey,
       _sendDirection = sendDirection,
       _receiveDirection = receiveDirection,
       _expiresAtMs = expiresAtMs;

  final String sessionId;
  final SecretKey _sendKey;
  final SecretKey _receiveKey;
  final int _sendDirection;
  final int _receiveDirection;
  final int _expiresAtMs;
  int _sendCounter = 0;
  int _highestReceiveCounter = -1;

  static Future<(CompanionSession, CompanionSession)> establish({
    required String sessionId,
    required List<int> sharedCredential,
    required List<int> initiatorNonce,
    required List<int> responderNonce,
    required int expiresAtMs,
  }) async {
    if (sharedCredential.length != 32 ||
        initiatorNonce.length != 32 ||
        responderNonce.length != 32 ||
        expiresAtMs < 0) {
      throw const CompanionProtocolException(
        'INVALID_SESSION_HANDSHAKE',
        'Session handshake material is invalid.',
      );
    }
    final salt = hashes.sha256.convert(<int>[
      ...initiatorNonce,
      ...responderNonce,
    ]).bytes;
    final keyMaterial = await Hkdf(hmac: Hmac.sha256(), outputLength: 64)
        .deriveKey(
          secretKey: SecretKey(sharedCredential),
          nonce: salt,
          info: utf8.encode('$companionMediaTransferSchema:$sessionId'),
        );
    final bytes = await keyMaterial.extractBytes();
    final initiatorSend = SecretKey(bytes.sublist(0, 32));
    final responderSend = SecretKey(bytes.sublist(32, 64));
    return (
      CompanionSession._(
        sessionId: sessionId,
        sendKey: initiatorSend,
        receiveKey: responderSend,
        sendDirection: 0x49325231,
        receiveDirection: 0x52324931,
        expiresAtMs: expiresAtMs,
      ),
      CompanionSession._(
        sessionId: sessionId,
        sendKey: responderSend,
        receiveKey: initiatorSend,
        sendDirection: 0x52324931,
        receiveDirection: 0x49325231,
        expiresAtMs: expiresAtMs,
      ),
    );
  }

  Future<String> seal({
    required CompanionMessageType type,
    required String messageId,
    required Map<String, Object?> payload,
    required int nowMs,
  }) async {
    _requireLive(nowMs);
    final envelope = CompanionEnvelope(
      type: type,
      messageId: messageId,
      sessionId: sessionId,
      counter: _sendCounter++,
      payload: payload,
    );
    final cleartext = utf8.encode(envelope.encode());
    final nonce = _nonce(envelope.counter, direction: _sendDirection);
    final aad = utf8.encode('$companionMediaTransferSchema:$sessionId');
    final box = await AesGcm.with256bits().encrypt(
      cleartext,
      secretKey: _sendKey,
      nonce: nonce,
      aad: aad,
    );
    return jsonEncode(<String, Object>{
      'schema': companionMediaTransferSchema,
      'sessionId': sessionId,
      'counter': envelope.counter,
      'ciphertext': base64Encode(box.cipherText),
      'mac': base64Encode(box.mac.bytes),
    });
  }

  Future<CompanionEnvelope> open({
    required String sealed,
    required int nowMs,
  }) async {
    _requireLive(nowMs);
    if (utf8.encode(sealed).length > companionMaximumMetadataBytes * 2) {
      throw const CompanionProtocolException(
        'SEALED_MESSAGE_TOO_LARGE',
        'Encrypted envelope exceeds the limit.',
      );
    }
    final Object? decoded;
    try {
      decoded = jsonDecode(sealed);
    } on FormatException {
      throw const CompanionProtocolException(
        'INVALID_SEALED_MESSAGE',
        'Encrypted envelope is malformed.',
      );
    }
    if (decoded is! Map) {
      throw const CompanionProtocolException(
        'INVALID_SEALED_MESSAGE',
        'Encrypted envelope is malformed.',
      );
    }
    final map = decoded.cast<String, Object?>();
    if (map.keys.toSet().difference(<String>{
          'schema',
          'sessionId',
          'counter',
          'ciphertext',
          'mac',
        }).isNotEmpty ||
        map.length != 5 ||
        map['schema'] != companionMediaTransferSchema ||
        map['sessionId'] != sessionId ||
        map['counter'] is! int ||
        map['ciphertext'] is! String ||
        map['mac'] is! String) {
      throw const CompanionProtocolException(
        'INVALID_SEALED_MESSAGE',
        'Encrypted envelope fields are invalid.',
      );
    }
    final counter = map['counter']! as int;
    if (counter <= _highestReceiveCounter) {
      throw const CompanionProtocolException(
        'REPLAY_REJECTED',
        'Message counter was already received.',
      );
    }
    final List<int> ciphertext;
    final List<int> mac;
    try {
      ciphertext = base64Decode(map['ciphertext']! as String);
      mac = base64Decode(map['mac']! as String);
    } on FormatException {
      throw const CompanionProtocolException(
        'INVALID_SEALED_MESSAGE',
        'Encrypted envelope payload is malformed.',
      );
    }
    final aad = utf8.encode('$companionMediaTransferSchema:$sessionId');
    final List<int> cleartext;
    try {
      cleartext = await AesGcm.with256bits().decrypt(
        SecretBox(
          ciphertext,
          nonce: _nonce(counter, direction: _receiveDirection),
          mac: Mac(mac),
        ),
        secretKey: _receiveKey,
        aad: aad,
      );
    } on SecretBoxAuthenticationError {
      throw const CompanionProtocolException(
        'AUTHENTICATION_FAILED',
        'Encrypted envelope authentication failed.',
      );
    }
    final envelope = CompanionEnvelope.decode(utf8.decode(cleartext));
    if (envelope.counter != counter || envelope.sessionId != sessionId) {
      throw const CompanionProtocolException(
        'COUNTER_MISMATCH',
        'Encrypted envelope counter does not match.',
      );
    }
    _highestReceiveCounter = counter;
    return envelope;
  }

  Future<Uint8List> sealBytes({
    required List<int> plaintext,
    required int nowMs,
  }) async {
    _requireLive(nowMs);
    if (plaintext.isEmpty || plaintext.length > companionMaximumChunkBytes) {
      throw const CompanionProtocolException(
        'INVALID_BINARY_PAYLOAD',
        'Binary payload is outside the chunk limit.',
      );
    }
    final counter = _sendCounter++;
    final aad = utf8.encode(
      '$companionMediaTransferSchema:$sessionId:binary:$counter',
    );
    final box = await AesGcm.with256bits().encrypt(
      plaintext,
      secretKey: _sendKey,
      nonce: _nonce(counter, direction: _sendDirection),
      aad: aad,
    );
    final header = ByteData(25);
    header.setUint8(0, 1);
    header.setUint64(1, counter, Endian.big);
    final output = BytesBuilder(copy: false)
      ..add(header.buffer.asUint8List(0, 9))
      ..add(box.mac.bytes)
      ..add(box.cipherText);
    return output.takeBytes();
  }

  Future<Uint8List> openBytes({
    required List<int> packet,
    required int nowMs,
  }) async {
    _requireLive(nowMs);
    if (packet.length < 26 ||
        packet.length > companionMaximumChunkBytes + 25 ||
        packet.first != 1) {
      throw const CompanionProtocolException(
        'INVALID_BINARY_PACKET',
        'Encrypted binary packet is malformed.',
      );
    }
    final header = ByteData.sublistView(Uint8List.fromList(packet), 0, 9);
    final counter = header.getUint64(1, Endian.big);
    if (counter <= _highestReceiveCounter) {
      throw const CompanionProtocolException(
        'REPLAY_REJECTED',
        'Binary packet counter was already received.',
      );
    }
    final aad = utf8.encode(
      '$companionMediaTransferSchema:$sessionId:binary:$counter',
    );
    final List<int> cleartext;
    try {
      cleartext = await AesGcm.with256bits().decrypt(
        SecretBox(
          packet.sublist(25),
          nonce: _nonce(counter, direction: _receiveDirection),
          mac: Mac(packet.sublist(9, 25)),
        ),
        secretKey: _receiveKey,
        aad: aad,
      );
    } on SecretBoxAuthenticationError {
      throw const CompanionProtocolException(
        'AUTHENTICATION_FAILED',
        'Encrypted binary authentication failed.',
      );
    }
    _highestReceiveCounter = counter;
    return Uint8List.fromList(cleartext);
  }

  void _requireLive(int nowMs) {
    if (nowMs > _expiresAtMs) {
      throw const CompanionProtocolException(
        'SESSION_EXPIRED',
        'Encrypted session has expired.',
      );
    }
  }

  Uint8List _nonce(int counter, {required int direction}) {
    if (counter < 0 || counter > 0x7fffffffffffffff) {
      throw const CompanionProtocolException(
        'INVALID_COUNTER',
        'Session counter is invalid.',
      );
    }
    final bytes = ByteData(12);
    bytes.setUint32(0, direction, Endian.big);
    bytes.setUint64(4, counter, Endian.big);
    return bytes.buffer.asUint8List();
  }
}

String companionFingerprint(List<int> publicKeyBytes) {
  final digest = hashes.sha256.convert(publicKeyBytes).bytes;
  return _base32(digest.sublist(0, 20));
}

String _base32(List<int> bytes) {
  const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';
  final output = StringBuffer();
  var value = 0;
  var bits = 0;
  for (final byte in bytes) {
    value = (value << 8) | byte;
    bits += 8;
    while (bits >= 5) {
      output.write(alphabet[(value >> (bits - 5)) & 31]);
      bits -= 5;
    }
  }
  if (bits > 0) output.write(alphabet[(value << (5 - bits)) & 31]);
  return output.toString();
}

File: apps/desktop-electron/src/main/domain/companion/companion_crypto.ts
import {
  createCipheriv,
  createDecipheriv,
  createHash,
  createPrivateKey,
  createPublicKey,
  diffieHellman,
  generateKeyPairSync,
  hkdfSync,
  type KeyObject,
  verify,
} from "node:crypto";

import {
  companionCapability,
  companionLimits,
  companionProtocol,
} from "../../../shared/contracts";

const initiatorDirection = 0x49325231;
const responderDirection = 0x52324931;
const messageTypes = new Set([
  "discovery",
  "pairingTranscript",
  "capability",
  "manifest",
  "chunk",
  "checkpoint",
  "receipt",
  "cancel",
  "error",
]);
const identifier = /^[a-zA-Z0-9][a-zA-Z0-9._:-]{0,127}$/;
const fingerprintPattern = /^[A-Z2-7]{20,64}$/;
const sha256Pattern = /^[a-f0-9]{64}$/;

export interface CompanionPairingTranscriptValue {
  schema: typeof companionProtocol;
  pairingId: string;
  initiatorDeviceId: string;
  initiatorFingerprint: string;
  initiatorEphemeralPublicKey: string;
  responderDeviceId: string;
  responderFingerprint: string;
  responderEphemeralPublicKey: string;
  shortCodeHash: string;
  expiresAtMs: number;
  capabilities: [typeof companionCapability];
}

export interface CompanionX25519KeyPair {
  publicKey: Buffer;
  privateKey: Buffer;
  destroy(): void;
}

export function generateCompanionX25519KeyPair(): CompanionX25519KeyPair {
  const pair = generateKeyPairSync("x25519");
  const privateDer = pair.privateKey.export({ format: "der", type: "pkcs8" });
  const privateBytes = Buffer.from(privateDer.subarray(-32));
  privateDer.fill(0);
  return {
    publicKey: rawPublicKey(pair.publicKey),
    privateKey: privateBytes,
    destroy: () => privateBytes.fill(0),
  };
}

export function companionX25519Secret(
  privateKey: Buffer,
  peerPublicKey: Buffer,
): Buffer {
  requireBytes(privateKey, 32, "pairingPrivateKey");
  requireBytes(peerPublicKey, 32, "peerEphemeralPublicKey");
  const secret = Buffer.from(
    diffieHellman({
      privateKey: x25519PrivateKey(privateKey),
      publicKey: x25519PublicKey(peerPublicKey),
    }),
  );
  if (secret.every((byte) => byte === 0)) {
    secret.fill(0);
    throw new CompanionCryptoError(
      "INVALID_PAIRING_KEY",
      "pairing shared secret is invalid",
    );
  }
  return secret;
}

export function deriveCompanionPairingCredential(
  secret: Buffer,
  pairingId: string,
  purpose: "temporary-channel" | "long-term-peer",
  transcriptHash?: Buffer,
): Buffer {
  requireBytes(secret, 32, "pairingSecret");
  requireIdentifier(pairingId, "pairingId");
  if (purpose === "long-term-peer") {
    requireBytes(transcriptHash, 32, "transcriptHash");
  }
  const salt = createHash("sha256")
    .update(companionProtocol)
    .update(":pairing:")
    .update(pairingId)
    .digest();
  return Buffer.from(
    hkdfSync(
      "sha256",
      secret,
      salt,
      Buffer.concat([
        Buffer.from(`${companionProtocol}:pairing:${purpose}`, "utf8"),
        transcriptHash ?? Buffer.alloc(0),
      ]),
      32,
    ),
  );
}

export function canonicalCompanionPairingTranscript(
  value: CompanionPairingTranscriptValue,
): Buffer {
  validatePairingTranscript(value);
  return Buffer.from(
    JSON.stringify({
      schema: companionProtocol,
      pairingId: value.pairingId,
      initiatorDeviceId: value.initiatorDeviceId,
      initiatorFingerprint: value.initiatorFingerprint,
      initiatorEphemeralPublicKey: value.initiatorEphemeralPublicKey,
      responderDeviceId: value.responderDeviceId,
      responderFingerprint: value.responderFingerprint,
      responderEphemeralPublicKey: value.responderEphemeralPublicKey,
      shortCodeHash: value.shortCodeHash,
      expiresAtMs: value.expiresAtMs,
      capabilities: value.capabilities,
    }),
    "utf8",
  );
}

export function verifyCompanionPairingSignature(
  transcript: CompanionPairingTranscriptValue,
  identityPublicKey: Buffer,
  signature: Buffer,
  expectedFingerprint: string,
): void {
  requireBytes(identityPublicKey, 32, "identityPublicKey");
  requireBytes(signature, 64, "identitySignature");
  if (companionFingerprint(identityPublicKey) !== expectedFingerprint) {
    throw new CompanionCryptoError(
      "PAIRING_FINGERPRINT_MISMATCH",
      "pairing identity fingerprint changed",
    );
  }
  if (
    !verify(
      null,
      canonicalCompanionPairingTranscript(transcript),
      ed25519PublicKey(identityPublicKey),
      signature,
    )
  ) {
    throw new CompanionCryptoError(
      "PAIRING_SIGNATURE_INVALID",
      "pairing signature is invalid",
    );
  }
}

export function companionFingerprint(publicKey: Buffer): string {
  requireBytes(publicKey, 32, "identityPublicKey");
  return base32(
    createHash("sha256").update(publicKey).digest().subarray(0, 20),
  );
}

export class CompanionCryptoError extends Error {
  constructor(
    readonly code: string,
    message: string,
  ) {
    super(message);
    this.name = "CompanionCryptoError";
  }
}

export interface CompanionCryptoEnvelope {
  schema: typeof companionProtocol;
  type: string;
  messageId: string;
  sessionId: string;
  counter: number;
  payload: Record<string, unknown>;
}

export class CompanionCryptoSession {
  private readonly sendKey: Buffer;
  private readonly receiveKey: Buffer;
  private readonly sendDirection: number;
  private readonly receiveDirection: number;
  private sendCounter = 0;
  private highestReceiveCounter = -1;

  constructor(
    private readonly options: {
      role: "initiator" | "responder";
      sessionId: string;
      sharedCredential: Buffer;
      initiatorNonce: Buffer;
      responderNonce: Buffer;
      expiresAtMs: number;
      nowMs?: () => number;
    },
  ) {
    requireIdentifier(options.sessionId, "sessionId");
    requireBytes(options.sharedCredential, 32, "sharedCredential");
    requireBytes(options.initiatorNonce, 32, "initiatorNonce");
    requireBytes(options.responderNonce, 32, "responderNonce");
    if (!Number.isSafeInteger(options.expiresAtMs) || options.expiresAtMs < 0) {
      throw new CompanionCryptoError(
        "INVALID_SESSION_HANDSHAKE",
        "invalid expiry",
      );
    }
    const salt = createHash("sha256")
      .update(options.initiatorNonce)
      .update(options.responderNonce)
      .digest();
    const material = Buffer.from(
      hkdfSync(
        "sha256",
        options.sharedCredential,
        salt,
        Buffer.from(`${companionProtocol}:${options.sessionId}`, "utf8"),
        64,
      ),
    );
    const initiatorSend = Buffer.from(material.subarray(0, 32));
    const responderSend = Buffer.from(material.subarray(32, 64));
    material.fill(0);
    if (options.role === "initiator") {
      this.sendKey = initiatorSend;
      this.receiveKey = responderSend;
      this.sendDirection = initiatorDirection;
      this.receiveDirection = responderDirection;
    } else {
      this.sendKey = responderSend;
      this.receiveKey = initiatorSend;
      this.sendDirection = responderDirection;
      this.receiveDirection = initiatorDirection;
    }
  }

  sealControl(
    type: string,
    messageId: string,
    payload: Record<string, unknown>,
  ): string {
    this.requireLive();
    if (!messageTypes.has(type)) {
      throw new CompanionCryptoError(
        "UNKNOWN_MESSAGE_TYPE",
        "unknown message type",
      );
    }
    requireIdentifier(messageId, "messageId");
    requireBoundedJson(payload, companionLimits.maximumMetadataBytes);
    const counter = this.nextSendCounter();
    const envelope: CompanionCryptoEnvelope = {
      schema: companionProtocol,
      type,
      messageId,
      sessionId: this.options.sessionId,
      counter,
      payload,
    };
    const cleartext = Buffer.from(JSON.stringify(envelope), "utf8");
    if (cleartext.length > companionLimits.maximumMetadataBytes) {
      throw new CompanionCryptoError(
        "METADATA_TOO_LARGE",
        "envelope exceeds limit",
      );
    }
    const encrypted = encrypt(
      this.sendKey,
      nonce(this.sendDirection, counter),
      cleartext,
      Buffer.from(`${companionProtocol}:${this.options.sessionId}`, "utf8"),
    );
    return JSON.stringify({
      schema: companionProtocol,
      sessionId: this.options.sessionId,
      counter,
      ciphertext: encrypted.ciphertext.toString("base64"),
      mac: encrypted.tag.toString("base64"),
    });
  }

  openControl(sealed: string): CompanionCryptoEnvelope {
    this.requireLive();
    if (
      Buffer.byteLength(sealed, "utf8") >
      companionLimits.maximumMetadataBytes * 2
    ) {
      throw new CompanionCryptoError(
        "SEALED_MESSAGE_TOO_LARGE",
        "sealed message exceeds limit",
      );
    }
    const wrapper = parseObject(sealed, "INVALID_SEALED_MESSAGE");
    exactKeys(wrapper, ["schema", "sessionId", "counter", "ciphertext", "mac"]);
    const counter = requireCounter(wrapper.counter);
    if (
      wrapper.schema !== companionProtocol ||
      wrapper.sessionId !== this.options.sessionId ||
      counter <= this.highestReceiveCounter
    ) {
      throw new CompanionCryptoError(
        counter <= this.highestReceiveCounter
          ? "REPLAY_REJECTED"
          : "INVALID_SEALED_MESSAGE",
        "sealed message identity or counter is invalid",
      );
    }
    const ciphertext = decodeCanonicalBase64(
      wrapper.ciphertext,
      undefined,
      "ciphertext",
    );
    const mac = decodeCanonicalBase64(wrapper.mac, 16, "mac");
    let cleartext: Buffer;
    try {
      cleartext = decrypt(
        this.receiveKey,
        nonce(this.receiveDirection, counter),
        ciphertext,
        mac,
        Buffer.from(`${companionProtocol}:${this.options.sessionId}`, "utf8"),
      );
    } catch {
      throw new CompanionCryptoError(
        "AUTHENTICATION_FAILED",
        "control authentication failed",
      );
    }
    if (cleartext.length > companionLimits.maximumMetadataBytes) {
      throw new CompanionCryptoError(
        "METADATA_TOO_LARGE",
        "envelope exceeds limit",
      );
    }
    const envelope = parseObject(cleartext.toString("utf8"), "INVALID_JSON");
    exactKeys(envelope, [
      "schema",
      "type",
      "messageId",
      "sessionId",
      "counter",
      "payload",
    ]);
    if (
      envelope.schema !== companionProtocol ||
      envelope.sessionId !== this.options.sessionId ||
      envelope.counter !== counter ||
      typeof envelope.type !== "string" ||
      !messageTypes.has(envelope.type) ||
      typeof envelope.messageId !== "string" ||
      !identifier.test(envelope.messageId) ||
      !isPlainObject(envelope.payload)
    ) {
      throw new CompanionCryptoError(
        "INVALID_ENVELOPE",
        "control envelope is invalid",
      );
    }
    requireBoundedJson(envelope.payload, companionLimits.maximumMetadataBytes);
    this.highestReceiveCounter = counter;
    return envelope as unknown as CompanionCryptoEnvelope;
  }

  sealBinary(plaintext: Buffer): Buffer {
    this.requireLive();
    if (
      plaintext.length < 1 ||
      plaintext.length > companionLimits.maximumChunkBytes
    ) {
      throw new CompanionCryptoError(
        "INVALID_BINARY_PAYLOAD",
        "binary payload is outside limit",
      );
    }
    const counter = this.nextSendCounter();
    const encrypted = encrypt(
      this.sendKey,
      nonce(this.sendDirection, counter),
      plaintext,
      Buffer.from(
        `${companionProtocol}:${this.options.sessionId}:binary:${counter}`,
        "utf8",
      ),
    );
    const packet = Buffer.allocUnsafe(25 + encrypted.ciphertext.length);
    packet[0] = 1;
    packet.writeBigUInt64BE(BigInt(counter), 1);
    encrypted.tag.copy(packet, 9);
    encrypted.ciphertext.copy(packet, 25);
    return packet;
  }

  openBinary(packet: Buffer): Buffer {
    this.requireLive();
    if (
      packet.length < 26 ||
      packet.length > companionLimits.maximumChunkBytes + 25 ||
      packet[0] !== 1
    ) {
      throw new CompanionCryptoError(
        "INVALID_BINARY_PACKET",
        "binary packet is malformed",
      );
    }
    const rawCounter = packet.readBigUInt64BE(1);
    if (rawCounter > BigInt(Number.MAX_SAFE_INTEGER)) {
      throw new CompanionCryptoError(
        "INVALID_COUNTER",
        "counter is outside safe range",
      );
    }
    const counter = Number(rawCounter);
    if (counter <= this.highestReceiveCounter) {
      throw new CompanionCryptoError(
        "REPLAY_REJECTED",
        "binary packet was already received",
      );
    }
    let cleartext: Buffer;
    try {
      cleartext = decrypt(
        this.receiveKey,
        nonce(this.receiveDirection, counter),
        packet.subarray(25),
        packet.subarray(9, 25),
        Buffer.from(
          `${companionProtocol}:${this.options.sessionId}:binary:${counter}`,
          "utf8",
        ),
      );
    } catch {
      throw new CompanionCryptoError(
        "AUTHENTICATION_FAILED",
        "binary authentication failed",
      );
    }
    this.highestReceiveCounter = counter;
    return cleartext;
  }

  destroy(): void {
    this.sendKey.fill(0);
    this.receiveKey.fill(0);
  }

  private nextSendCounter(): number {
    if (this.sendCounter > Number.MAX_SAFE_INTEGER) {
      throw new CompanionCryptoError("INVALID_COUNTER", "counter exhausted");
    }
    return this.sendCounter++;
  }

  private requireLive(): void {
    if ((this.options.nowMs ?? Date.now)() > this.options.expiresAtMs) {
      throw new CompanionCryptoError(
        "SESSION_EXPIRED",
        "encrypted session expired",
      );
    }
  }
}

export class CompanionReceiverCryptoSession extends CompanionCryptoSession {
  constructor(
    options: Omit<
      ConstructorParameters<typeof CompanionCryptoSession>[0],
      "role"
    >,
  ) {
    super({ ...options, role: "responder" });
  }
}

export function decodeCanonicalBase64(
  value: unknown,
  expectedBytes?: number,
  field = "value",
): Buffer {
  if (
    typeof value !== "string" ||
    value.length === 0 ||
    value.length % 4 !== 0 ||
    !/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/.test(
      value,
    )
  ) {
    throw new CompanionCryptoError(
      "INVALID_BASE64",
      `${field} is not canonical base64`,
    );
  }
  const decoded = Buffer.from(value, "base64");
  if (
    decoded.toString("base64") !== value ||
    (expectedBytes !== undefined && decoded.length !== expectedBytes)
  ) {
    throw new CompanionCryptoError(
      "INVALID_BASE64",
      `${field} has invalid length`,
    );
  }
  return decoded;
}

function encrypt(key: Buffer, iv: Buffer, cleartext: Buffer, aad: Buffer) {
  const cipher = createCipheriv("aes-256-gcm", key, iv);
  cipher.setAAD(aad);
  return {
    ciphertext: Buffer.concat([cipher.update(cleartext), cipher.final()]),
    tag: cipher.getAuthTag(),
  };
}

function decrypt(
  key: Buffer,
  iv: Buffer,
  ciphertext: Buffer,
  tag: Buffer,
  aad: Buffer,
): Buffer {
  const decipher = createDecipheriv("aes-256-gcm", key, iv);
  decipher.setAAD(aad);
  decipher.setAuthTag(tag);
  return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
}

function nonce(direction: number, counter: number): Buffer {
  if (!Number.isSafeInteger(counter) || counter < 0) {
    throw new CompanionCryptoError("INVALID_COUNTER", "counter is invalid");
  }
  const output = Buffer.alloc(12);
  output.writeUInt32BE(direction, 0);
  output.writeBigUInt64BE(BigInt(counter), 4);
  return output;
}

function parseObject(encoded: string, code: string): Record<string, unknown> {
  let decoded: unknown;
  try {
    decoded = JSON.parse(encoded);
  } catch {
    throw new CompanionCryptoError(code, "JSON is malformed");
  }
  if (!isPlainObject(decoded))
    throw new CompanionCryptoError(code, "JSON object required");
  return decoded;
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function exactKeys(
  value: Record<string, unknown>,
  expected: readonly string[],
): void {
  const keys = Object.keys(value);
  if (
    keys.length !== expected.length ||
    expected.some((key) => !Object.hasOwn(value, key))
  ) {
    throw new CompanionCryptoError(
      "INVALID_FIELDS",
      "missing or unknown fields",
    );
  }
}

function requireCounter(value: unknown): number {
  if (!Number.isSafeInteger(value) || Number(value) < 0) {
    throw new CompanionCryptoError("INVALID_COUNTER", "counter is invalid");
  }
  return Number(value);
}

function requireIdentifier(value: string, field: string): void {
  if (!identifier.test(value)) {
    throw new CompanionCryptoError("INVALID_IDENTIFIER", `${field} is invalid`);
  }
}

function requireBytes(
  value: Buffer | undefined,
  length: number,
  field: string,
): asserts value is Buffer {
  if (!Buffer.isBuffer(value) || value.length !== length) {
    throw new CompanionCryptoError(
      "INVALID_SESSION_HANDSHAKE",
      `${field} is invalid`,
    );
  }
}

function requireBoundedJson(
  value: Record<string, unknown>,
  maximum: number,
): void {
  let encoded: string;
  try {
    encoded = JSON.stringify(value);
  } catch {
    throw new CompanionCryptoError("INVALID_JSON", "value is not serializable");
  }
  if (Buffer.byteLength(encoded, "utf8") > maximum) {
    throw new CompanionCryptoError(
      "METADATA_TOO_LARGE",
      "metadata exceeds limit",
    );
  }
}

function validatePairingTranscript(
  value: CompanionPairingTranscriptValue,
): void {
  if (
    value.schema !== companionProtocol ||
    !identifier.test(value.pairingId) ||
    !identifier.test(value.initiatorDeviceId) ||
    !identifier.test(value.responderDeviceId) ||
    !fingerprintPattern.test(value.initiatorFingerprint) ||
    !fingerprintPattern.test(value.responderFingerprint) ||
    !sha256Pattern.test(value.shortCodeHash) ||
    !Number.isSafeInteger(value.expiresAtMs) ||
    value.expiresAtMs < 0 ||
    value.capabilities.length !== 1 ||
    value.capabilities[0] !== companionCapability
  ) {
    throw new CompanionCryptoError(
      "INVALID_PAIRING_TRANSCRIPT",
      "pairing transcript is invalid",
    );
  }
  decodeCanonicalBase64(
    value.initiatorEphemeralPublicKey,
    32,
    "initiatorEphemeralPublicKey",
  );
  decodeCanonicalBase64(
    value.responderEphemeralPublicKey,
    32,
    "responderEphemeralPublicKey",
  );
}

function x25519PrivateKey(raw: Buffer): KeyObject {
  return createPrivateKey({
    key: Buffer.concat([
      Buffer.from("302e020100300506032b656e04220420", "hex"),
      raw,
    ]),
    format: "der",
    type: "pkcs8",
  });
}

function x25519PublicKey(raw: Buffer): KeyObject {
  return createPublicKey({
    key: Buffer.concat([Buffer.from("302a300506032b656e032100", "hex"), raw]),
    format: "der",
    type: "spki",
  });
}

function ed25519PublicKey(raw: Buffer): KeyObject {
  return createPublicKey({
    key: Buffer.concat([Buffer.from("302a300506032b6570032100", "hex"), raw]),
    format: "der",
    type: "spki",
  });
}

function rawPublicKey(key: KeyObject): Buffer {
  const der = key.export({ format: "der", type: "spki" });
  if (der.length !== 44) {
    throw new CompanionCryptoError(
      "INVALID_PAIRING_KEY",
      "pairing public key is invalid",
    );
  }
  return Buffer.from(der.subarray(-32));
}

function base32(bytes: Buffer): string {
  const alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
  let output = "";
  let accumulator = 0;
  let bits = 0;
  for (const byte of bytes) {
    accumulator = (accumulator << 8) | byte;
    bits += 8;
    while (bits >= 5) {
      bits -= 5;
      output += alphabet[(accumulator >>> bits) & 31];
    }
  }
  if (bits > 0) output += alphabet[(accumulator << (5 - bits)) & 31];
  return output;
}
